bzt/modules/apachebench.py

```diff
@@ -0,0 +1,204 @@
+        request = requests[0]
+
+        # add request-specific headers
+        for header in request.headers:
+            for key, val in iteritems(header):
+                args += ['-H', "%s: %s" % (key, val)]
+
+        args += ['-m', request.method]
+
+        if request.timeout is not None:
+            timeout = int(ceil(dehumanize_time(request.timeout)))
+            args += ['-s', str(timeout)]
+
+        args += [request.url]
+
+        self.start_time = time.time()
+        self.process = shell_exec(args, stdout=self.__out, stderr=self.__err)
+
+    def check(self):
+        ret_code = self.process.poll()
+        if ret_code is None:
+            return False
+        self.log.info("ApacheBench tool exit code: %s", ret_code)
+        if ret_code != 0:
+            raise RuntimeError("ApacheBench tool exited with non-zero code")
+        return True
+
+    def shutdown(self):
+        """
+        If tool is still running - let's stop it.
+        """
+        shutdown_process(self.process, self.log)
+        if self.__out and not self.__out.closed:
+            self.__out.close()
+        if self.__err and not self.__err.closed:
+            self.__err.close()
+
+    def _check_installed(self):
+        tool_path = self.settings.get('path', 'ab')
+        ab = ApacheBench(tool_path, self.log)
+        if not ab.check_if_installed():
+            raise RuntimeError("You must install ApacheBench tool at first")
+        return tool_path
+
+
+class TSVDataReader(ResultsReader):
+    def __init__(self, filename, parent_logger):
+        super(TSVDataReader, self).__init__()
+        self.log = parent_logger.getChild(self.__class__.__name__)
+        self.filename = filename
+        self.fds = None
+        self.concurrency = None
+
+    def _calculate_datapoints(self, final_pass=False):
+        for point in super(TSVDataReader, self)._calculate_datapoints(final_pass):
+            yield point
+
+    def __open_fds(self):
+        if not path.isfile(self.filename):
+            self.log.debug("File not appeared yet")
+            return False
+
+        if not path.getsize(self.filename):
+            self.log.debug("File is empty: %s", self.filename)
+            return False
+
+        if not self.fds:
+            self.fds = open(self.filename)
+
+        return True
+
+    def _read(self, last_pass=False):
+        while not self.fds and not self.__open_fds():
+            self.log.debug("No data to start reading yet")
+            yield None
+        if last_pass:
+            lines = self.fds.readlines()  # unlimited
+            self.fds.close()
+        else:
+            lines = self.fds.readlines(1024 * 1024)  # 1MB limit to read    git
+
+        for line in lines:
+            log_vals = [val.strip() for val in line.split('\t')]
+
+            if all(val.isalpha() for val in log_vals):
+                continue
+
+            _latency = 0
+            _error = None
+            _url = None
```

Comment:
we know the URL. in fact, it is "label" expected here. and we have the label as part of request